### ISSUE TITLE
Please include the C client in this package instead of downloading it on every rebuild

### ISSUE THREAD

**Issue by revington** [aerospike/aerospike-client-nodejs#92]
Otherwise you are forcing me to download this package on every deploy. If I can not download the package I can not deploy :(
For production environmensts you usually do: 
```
$ npm install --ignore-scripts
$ git add node_modules
$ git commit -m 'deps'
```
Then in production just build your stuff. No downloads!
```
$ npm rebuild
```
With this module that can not be done because on `npm rebuild` it downloads the C client from aerospike. **Therefore, if the aerospike site is down I can not deploy.**

http://www.letscodejavascript.com/v3/blog/2014/03/the_npm_debacle

I discovered that becasue today my travis test are failing due a connectivity error with aeropspike site.

```
$ wget -O aerospike-server.tgz http://aerospike.com/download/server/latest/artifact/tgz
--2015-09-22 07:25:07--  http://aerospike.com/download/server/latest/artifact/tgz
Resolving aerospike.com (aerospike.com)... 184.168.221.29
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:26:11--  (try: 2)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:27:16--  (try: 3)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:28:22--  (try: 4)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:29:29--  (try: 5)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:30:37--  (try: 6)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:31:46--  (try: 7)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:32:56--  (try: 8)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:34:07--  (try: 9)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:35:19--  (try:10)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:36:32--  (try:11)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:37:45--  (try:12)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... failed: Connection timed out.
Retrying.
--2015-09-22 07:38:58--  (try:13)  http://aerospike.com/download/server/latest/artifact/tgz
Connecting to aerospike.com (aerospike.com)|184.168.221.29|:80... 
```

**Comment by ApsOps** [comment]
This is a huge problem that has been troubling us for weeks now :(

**Comment by ApsOps** [comment]
@jhecking Hi. Can we please do something about this? :)

**Comment by jhecking** [comment]
Hi @ApsOps, I'm looking into this issue right now actually. :) Trying to determine what are the different options for resolving this.

**Comment by ApsOps** [comment]
Great! Thanks.

**Comment by jhecking** [comment]
I really don't like the idea of including the pre-compiled Aerospike C/C++ client library in the Aerospike npm package. We would have to include static libraries (*.a) for all supported platforms. That's 7 different versions of the library at the moment, each about 3-5 MB in size. (Plus header files and system Lua scripts - but those are relatively small and platform independent.)

What if I rework the `aerospike-client-c.sh` such that it can be called as a separate npm script and can take an optional target platform parameter. Your build/deployment procedure would then look something like this:

    $ export target_platform=ubuntu14
    $ npm install --ignore-scripts
    $ npm explore aerospike -- npm run fetch_binaries ${target_platform}
    $ git add node_modules
    $ git commit -m 'deps'

On prod:

    $ npm rebuild

`npm rebuild` would detect that the binaries are already present and just compile the Aerospike Node.js add-on using `node-gyp`.

Would something like this work for you?

**Comment by ApsOps** [comment]
@jhecking In some of the environments, like AWS Elastic Beanstalk for instance, users don't have much control in adding custom npm commands.

Would it be better if I can give another URL (perhaps a self-hosted version?) for C client via, say, an environment variable? Or maybe add "retry" logic in the `aerospike-client-c.sh` script?

**Comment by jhecking** [comment]
@ApsOps, I was referring to @revington's initial report in how he is deploying his apps. It was wrong to assume that you are following the same model.

Another idea I have been playing with is to package up the binaries as separate npm package - one per platform. But I still need to figure out a few details to see if this is even a viable option. Managing the package dependencies would be a problem for one. But I could always fall back to fetching the binaries at install time if the right binary package is not installed.

But I'm open for other ideas as well.

**Comment by revington** [comment]
I bundle apps by:
$ rm -rf node_modules
$ npm install --production --ignore-scripts
$ git add node_modules
...
In the production machine i run
$ npm rebuild

The problem with the aerospike module is that it should decide which c
client to download at "rebuild time".

My feeling is that the best practice will be to create
aerospike-client-{windows,Linux..} modules. So you install aerospike client
with:

$ npm i --save aerospike aerospike-client-X

The former installs this module, the later installs the client.

I'm sorry for a late reply, I'm on holidays and no computer so I'm
answering with my phone which (due to fat fingers and lack of ability) is a
pain in the ***

**Comment by revington** [comment]
@jhecking what do you think about my proposal?

**Comment by revington** [comment]
Today i've bundled the aerospike c client with the other modules. More or less:
```
$ rm -rf node_modules
$ npm i --production --ignore-scripts
$ git add -f node_modules
$ npm rebuild # this will download and compile the aerospike c client
$ rm node_modules/aerospike/aerospike-client-c/package/usr/lib/libaerospike.so
$ git add -f node_modules/aerospike/aerospike-client-c
$ git commit -m 'module bundle'
```
Then, in production I'll just `npm rebuild`, aerospike-client-c will no longer get downloaded.

**Comment by jhecking** [comment]
@revington, sorry for not responding earlier. I haven't made much progress on this yet since I've been busy with another project. (I maintain Aerospike's Ruby client as well.) But I'll get back to this soon.

In order to avoid the package download when you rebuild the package, you just need to ensure the `aerospike-client-c.tgz` archive can be found at `aerospike-client-c/package/`. It contains all the platform specific binaries so rebuilding the package should work on all supported platforms.

I.e. instead of removing just `libaerospike.so` you can remove everything but `aerospike-client-c.tgz`:

```
$ (cd node_modules/aerospike/aerospike-client-c/; rm -r include lib lua package/usr package/aerospike-client-c-libuv-*)
$ find node_modules/aerospike/aerospike-client-c
node_modules/aerospike/aerospike-client-c
node_modules/aerospike/aerospike-client-c/package
node_modules/aerospike/aerospike-client-c/package/aerospike-client-c.tgz
$ npm rebuild

> aerospike@2.0.4 install /Users/jhecking/Projects/aerospike/aerospike-92/node_modules/aerospike
> node-gyp rebuild

  TOUCH binding_gyp_aerospike_client_c_target_run_scripts_aerospike_client_c_sh.intermediate
  ACTION binding_gyp_aerospike_client_c_target_run_scripts_aerospike_client_c_sh binding_gyp_aerospike_client_c_target_run_scripts_aerospike_client_c_sh.intermediate
warning: 'aerospike-client-c/package/aerospike-client-c.tgz' file exists.
warning:
warning: We will be using this package, rather than downloading a new package.
warning: If you would like to download a new package, then please remove.
warning: 'aerospike-client.tgz' from this directory.
warning:
info: extracting 'aerospike-client-c-4.0.4.mac.x86_64/aerospike-client-c-libuv-devel-4.0.4.pkg' from 'aerospike-client-c.tgz'
info: extracting files from 'aerospike-client-c-4.0.4.mac.x86_64/aerospike-client-c-libuv-devel-4.0.4.pkg'
9804 blocks

CHECK
   [✓] /Users/jhecking/Projects/aerospike/aerospike-92/node_modules/aerospike/aerospike-client-c/package/usr/local/lib/libaerospike.a
   [✓] /Users/jhecking/Projects/aerospike/aerospike-92/node_modules/aerospike/aerospike-client-c/package/usr/local/include/aerospike/aerospike.h
   [✓] /Users/jhecking/Projects/aerospike/aerospike-92/node_modules/aerospike/aerospike-client-c/package/usr/local/aerospike/client/sys/udf/lua/aerospike.lua

...
aerospike@2.0.4 /Users/jhecking/Projects/aerospike/aerospike-92/node_modules/aerospike
nan@2.3.3 /Users/jhecking/Projects/aerospike/aerospike-92/node_modules/nan
```

**Comment by revington** [comment]
@jhecking thank you very much, that is even better!

**Comment by jhecking** [comment]
Correction: `aerospike-client-c.tgz` _is_ platform dependent. It only contains the binaries for one specific platform. (Not sure what I was thinking earlier...) So you still need to make sure that the platform on which you execute `npm rebuild` to download this package is the same platform as that to which you will deploy.

**Comment by jhecking** [comment]
> Would it be better if I can give another URL (perhaps a self-hosted version?) for C client via, say, an environment variable? Or maybe add "retry" logic in the aerospike-client-c.sh script?

@ApsOps, these are all good suggestions and I will look into them. As a first step I'm planning to host the package downloads in Amazon S3 and use CloudFront as a CDN. That should make the downloads faster and more reliable.

**Comment by ApsOps** [comment]
Thanks @jhecking. Moving package to S3 is definitely a good first step.

**Comment by revington** [comment]
Just an update, the node_modules/aerospike/build should be deleted too. I have created this bundle task for package.json https://gist.github.com/revington/426c9d64afceb20cf92242d787c3a3ca#file-package-json-L4
Just run `npm run-script bundle`

**Comment by jhecking** [comment]
I am going to close this as wontfix for now.

There are a number of improvements to the C client resolution process in the just released v2.3.0 build. Instead of fetching the large bundle file (30+ MB) the npm install script now fetches a much smaller package (~2 MB). This file is downloaded via a CDN (AWS CloudFront) so www.aerospike.com is no longer a direct dependency. (It still serves as a host of origin for the files served via the CDN; but in most cases the files should be cached by CF.)

There are currently no plans to include the C client binaries in the npm package itself.

**Comment by ApsOps** [comment]
@jhecking This makes me sad. It just broke our production deploys. `v2.3.0` doesn't install on Amazon Linux, but `v2.2.0` does.

```bash
[xxx]$ npm install aerospike@2.3.0
-
> aerospike@2.3.0 install /var/lib/pipeline/node_modules/aerospike
> node-gyp rebuild

make: Entering directory `/var/lib/pipeline/node_modules/aerospike/build'
  ACTION binding_gyp_aerospike_client_c_target_run_scripts_aerospike_client_c_sh aerospike-client-c/lib/libaerospike.a
info: downloading 'http://artifacts.aerospike.com/aerospike-client-c/4.0.7/' to '/var/lib/pipeline/node_modules/aerospike/aerospike-client-c/package/'
scripts/aerospike-client-c.sh: line 57: /var/lib/pipeline/node_modules/aerospike/aerospike-client-c/package/: Is a directory
error: Unable to download package from 'http://artifacts.aerospike.com/aerospike-client-c/4.0.7/'
make: *** [aerospike-client-c/lib/libaerospike.a] Error 1
make: Leaving directory `/var/lib/pipeline/node_modules/aerospike/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/.nodenv/versions/4.2.3/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Linux 3.14.35-28.38.amzn1.x86_64
gyp ERR! command "/usr/local/.nodenv/versions/4.2.3/bin/node" "/usr/local/.nodenv/versions/4.2.3/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /var/lib/pipeline/node_modules/aerospike
gyp ERR! node -v v4.2.3
gyp ERR! node-gyp -v v3.0.3
gyp ERR! not ok
npm ERR! Linux 3.14.35-28.38.amzn1.x86_64
npm ERR! argv "/usr/local/.nodenv/versions/4.2.3/bin/node" "/usr/local/.nodenv/versions/4.2.3/bin/npm" "install" "aerospike@2.3.0"
npm ERR! node v4.2.3
npm ERR! npm  v2.14.7
npm ERR! code ELIFECYCLE

npm ERR! aerospike@2.3.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the aerospike@2.3.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the aerospike package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls aerospike
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /var/lib/pipeline/npm-debug.log
```

**Comment by jhecking** [comment]
@ApsOps, my bad. I tested this release on more than half a dozen different distros, but did not include Amazon Linux. :-( Fortunately this is easy to fix and I should have a new 2.3.1 package out soon.
